integration/csi/alluxio/nodeserver.go

```diff
@@ -0,0 +1,133 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/golang/glog"
+	csicommon "github.com/kubernetes-csi/drivers/pkg/csi-common"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	mount "k8s.io/mount-utils"
+)
+
+type nodeServer struct {
+	nodeId string
+	*csicommon.DefaultNodeServer
+}
+
+func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
+	targetPath := req.GetTargetPath()
+
+	notMnt, err := mount.New("").IsLikelyNotMountPoint(targetPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(targetPath, 0750); err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
+			}
+			notMnt = true
+		} else {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	if !notMnt {
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
+
+	mountOptions := req.GetVolumeCapability().GetMount().GetMountFlags()
+	if req.GetReadonly() {
+		mountOptions = append(mountOptions, "ro")
+	}
+
+	/*
+	   https://docs.alluxio.io/os/user/edge/en/api/POSIX-API.html
+	   https://github.com/Alluxio/alluxio/blob/master/integration/fuse/bin/alluxio-fuse
+	*/
+
+	alluxioPath := req.GetVolumeContext()["alluxioPath"]
+	if alluxioPath == "" {
+		alluxioPath = "/"
+	}
+
+	args := []string{"mount"}
+	if len(mountOptions) > 0 {
+		args = append(args, "-o", strings.Join(mountOptions, ","))
+	}
+	args = append(args, targetPath, alluxioPath)
+	command := exec.Command("/opt/alluxio/integration/fuse/bin/alluxio-fuse", args...)
+
+	masterConfig := ""
+	if masterHost, ok := req.GetVolumeContext()["alluxio.master.hostname"]; ok {
```

Comment:
How to specify a `alluxio.master.hostname`?

Reply:
Under `parameters` filed of storage class
such as:
```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: alluxio
provisioner: alluxio
parameters:
  alluxio.master.hostname:  master_ip
volumeBindingMode: Immediate
```
But this can only be set in `javaOptions` filed. Add this option here just for backward compatibility